test/integration/xds_integration_test.cc

```diff
@@ -72,6 +73,355 @@ TEST_P(XdsIntegrationTestTypedStruct, RouterRequestAndResponseWithBodyNoBuffer)
+    std::string tls_inspector_config = ConfigHelper::tlsInspectorFilter();
+    config_helper_.addListenerFilter(tls_inspector_config);
+    config_helper_.addSslConfig();
+    config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto* filter_chain_0 =
+          bootstrap.mutable_static_resources()->mutable_listeners(0)->mutable_filter_chains(0);
+      *filter_chain_0->mutable_filter_chain_match()->mutable_application_protocols()->Add() =
+          "alpn0";
+      auto* filter_chain_1 = bootstrap.mutable_static_resources()
+                                 ->mutable_listeners(0)
+                                 ->mutable_filter_chains()
+                                 ->Add();
+      filter_chain_1->MergeFrom(*filter_chain_0);
+
+      // filter chain 1
+      // alpn1, route to cluster_1
+      *filter_chain_1->mutable_filter_chain_match()->mutable_application_protocols(0) = "alpn1";
+
+      auto* config_blob = filter_chain_1->mutable_filters(0)->mutable_typed_config();
+
+      ASSERT_TRUE(config_blob->Is<envoy::extensions::filters::network::http_connection_manager::v3::
+                                      HttpConnectionManager>());
+      auto hcm_config = MessageUtil::anyConvert<
+          envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager>(
+          *config_blob);
+      hcm_config.mutable_route_config()
+          ->mutable_virtual_hosts(0)
+          ->mutable_routes(0)
+          ->mutable_route()
+          ->set_cluster("cluster_1");
+      config_blob->PackFrom(hcm_config);
+      bootstrap.mutable_static_resources()->mutable_clusters()->Add()->MergeFrom(
+          *bootstrap.mutable_static_resources()->mutable_clusters(0));
+      bootstrap.mutable_static_resources()->mutable_clusters(1)->set_name("cluster_1");
+    });
+
+    BaseIntegrationTest::initialize();
+
+    context_manager_ =
+        std::make_unique<Extensions::TransportSockets::Tls::ContextManagerImpl>(timeSystem());
+    context_ = Ssl::createClientSslTransportSocketFactory({}, *context_manager_, *api_);
+  }
+
+  std::unique_ptr<SslClient> connect(const std::string& alpn) {
+    Network::Address::InstanceConstSharedPtr address =
+        Ssl::getSslAddress(version_, lookupPort("http"));
+    auto ssl_conn = dispatcher_->createClientConnection(
+        address, Network::Address::InstanceConstSharedPtr(),
+        context_->createTransportSocket(std::make_shared<Network::TransportSocketOptionsImpl>(
+            absl::string_view(""), std::vector<std::string>(), std::vector<std::string>{alpn})),
+        nullptr);
+    return std::make_unique<SslClient>(std::move(ssl_conn), *dispatcher_);
+  }
+
+  std::unique_ptr<Ssl::ContextManager> context_manager_;
+  Network::TransportSocketFactoryPtr context_;
+  testing::NiceMock<Secret::MockSecretManager> secret_manager_;
+};
+
+TEST_P(LdsInplaceUpdateHttpIntegrationTest, ReloadConfigAddingFilterChainWip) {
```

Comment:
Add a '//' comment above each test explaining what each test does. These tests are very complicated and need some explanation.